Sming/Components/ssl/include/Network/Ssl/Fingerprints.h

```diff
@@ -10,124 +10,70 @@
 
 #pragma once
 
-#include "Crypto.h"
-#include <sming_attr.h>
+#include <Crypto/Sha1.h>
+#include <Crypto/Sha256.h>
 
 namespace Ssl
 {
 /**
- * @brief SSL Certificate fingerprint type
- */
-enum FingerprintType {
-	/** @brief Fingerprint based on the SHA1 value of the certificate
-	 * 	@note The SHA1 hash of the entire certificate. This changes on each certificate renewal so needs
-	 * 	to be updated every time the remote server updates its certificate.
-	 * 	Advantages: Takes less time to verify than SHA256
-	 * 	Disadvantages: Likely to change periodically
-	 */
-	eSFT_CertSha1 = 0,
-
-	/**	@brief Fingerprint based on the SHA256 value of the Public Key Subject in the certificate
-	 * 	@note For HTTP public key pinning (RFC7469), the SHA-256 hash of the Subject Public Key Info
-	 * 	(which usually only changes when the public key changes) is used.
-	 * 	Advantages: Doesn't change frequently
-	 * 	Disadvantages: Takes more time (in ms) to verify.
-	 */
-	eSFT_PkSha256,
-};
-
-/** @brief Contains SSL fingerprint data
- *  @ingroup ssl
+ * @brief Various types of fingerprint
+ *
+ * This structure provides a convenient way to define unique types for each fingerprint.
  *
- *  Use within application's SSL initialisation callback to simplify setting standard
- *  fingerprints for validation. For example:
+ * It's not intended to actually be used as a union, that just implies we're picking one of them,
+ * not all of them.
  *
- *		void mySslInitCallback(Ssl::Session& session, HttpRequest& request)
- *		{
- *			static const uint8_t sha1Fingerprint[] PROGMEM = {
- *				0x15, 0x9A, 0x76, 0xC5, 0xAE, 0xF4, 0x90, 0x15, 0x79, 0xE6,
- *				0xA4, 0x99, 0x96, 0xC1, 0xD6, 0xA1, 0xD9, 0x3B, 0x07, 0x43
- *			};
- *			Ssl::Fingerprints fingerprints;
- *			fingerprints.setSha1_P(sha1Fingerprint, sizeof(sha1Fingerprint));
- *			session.validators.add(fingerprints);
- *		}
+ * For example:
  *
+ * 		Fingerprint::Cert::Sha1 fingerprint = { ... };
  */
-struct Fingerprints {
-	const uint8_t* certSha1 = nullptr; ///< certificate SHA1 fingerprint
-	const uint8_t* pkSha256 = nullptr; ///< public key SHA256 fingerprint
-
-	~Fingerprints()
-	{
-		free();
-	}
-
-	void free();
-
-	/** @brief Set the SHA1 fingerprint
-	 *  @param cert data to copy
-	 *  @param length for checking
-	 *  @retval bool false on length check or allocation failure
-	 */
-	bool setSha1(const uint8_t* cert, unsigned length)
-	{
-		return setValue(certSha1, SHA1_SIZE, cert, length);
-	}
-
-	/** @brief Make copy of SHA1 certificate from data stored in flash */
-	bool setSha1_P(const uint8_t* cert, unsigned length)
-	{
-		return setValue(certSha1, SHA1_SIZE, cert, length);
-	}
-
-	/** @brief Set the SHA256 fingerprint
-	 *  @param cert data to copy
-	 *  @param length for checking
-	 *  @retval bool false on length check or allocation failure
+union Fingerprint {
```

Comment:
According to the documentation comments above, it appears the same could have been achieved by using nested Namespaces, without the distraction created by the union and the restrictions implied on the contained types.

Reply:
I started with namespaces but we don't get nested formatting... I'm undecided about having multiple `Certificate::getFingerprint` methods, we could replace that with a single virtual method which takes a type enum plus Fingerprint union.

I'll take another look at this; won't affect the public interface, just the connection between the validator and the certificate.

Reply:
Adding fingerprint type enum back also gives us an efficient way for ValidatorList to track which fingerprint validators have been added. For BearSSL, we then only generate those fingerprints. For example, there's no point in validating both Cert::SHA1 and Cert::SHA256, we'd only ask for one of those. And if Pki::SHA256 isn't needed, we don't calculate it. That should speed things up considerably.